File: backend/src/main/java/com/flightmanagement/service/EmployeeService.java
package com.flightmanagement.service;

import com.flightmanagement.dto.EmployeeDto;

import java.util.List;

/**
 * Employee Service Interface
 * Provides comprehensive employee management functionality
 *
 * Last updated: 2025-06-11 07:46:23 UTC by thinh0704hcm
 */
public interface EmployeeService {

    // ===============================
    // BASIC CRUD OPERATIONS
    // ===============================

    /**
     * Get all active employees
     * @return List of all active employee DTOs
     */
    List<EmployeeDto> getAllEmployees();

    /**
     * Get employee by ID
     * @param employeeId Employee ID
     * @return Employee DTO if found
     * @throws RuntimeException if employee not found
     */
    EmployeeDto getEmployeeById(Integer employeeId);

    /**
     * Get current logged-in employee
     * @return Current employee DTO
     * @throws RuntimeException if no employee is logged in
     */
    EmployeeDto getCurrentEmployee(String accountName);

    /**
     * Update employee information
     * @param employeeId Employee ID to update
     * @param updateRequest Update data
     * @return Updated employee DTO
     * @throws RuntimeException if employee not found or update fails
     */
    EmployeeDto updateEmployee(Integer employeeId, EmployeeDto updateRequest);

    /**
     * Delete employee (soft delete)
     * @param employeeId Employee ID to delete
     * @throws RuntimeException if employee not found
     */
    void deleteEmployee(Integer employeeId);

    /**
     * Update an employee's role
     * @param id
     * @param newRole
     * @return
     */
    EmployeeDto updateRole(Integer id, Integer newRole);

    // ===============================
    // EMPLOYEE STATUS MANAGEMENT
    // ===============================

    /**
     * Activate employee account
     * @param employeeId Employee ID to activate
     * @return Updated employee DTO
     * @throws RuntimeException if employee not found
     */
    EmployeeDto activateEmployee(Integer employeeId);

    /**
     * Deactivate employee account
     * @param employeeId Employee ID to deactivate
     * @return Updated employee DTO
     * @throws RuntimeException if employee not found
     */
    EmployeeDto deactivateEmployee(Integer employeeId);

    /**
     * Check if employee is active
     * @param employeeId Employee ID to check
     * @return true if employee is active, false otherwise
     */
    boolean isEmployeeActive(Integer employeeId);
}

File: frontend/src/services/employeeService.ts
import { apiClient } from "./api";
import { Employee, UpdateEmployeeRequest } from "../models";
import { API_URL } from "./config";
import type { ApiResponse } from "../models/ApiResponse";
import type { PaginatedResponse } from "../models";

class EmployeeService {
  private readonly baseUrl = API_URL.EMPLOYEES;

  async getCurrentEmployee(): Promise<ApiResponse<Employee>> {
    return apiClient.get(`${this.baseUrl}/current`);
  }

  async getAllEmployees(): Promise<ApiResponse<Employee[]>> {
    return apiClient.get(this.baseUrl);
  }

  async getAllEmployeesPaged(
    page: number = 0,
    size: number = 10
  ): Promise<ApiResponse<PaginatedResponse<Employee>>> {
    return apiClient.get(`${this.baseUrl}?page=${page}&size=${size}`);
  }

  async getEmployeeById(id: number): Promise<ApiResponse<Employee>> {
    return apiClient.get(`${this.baseUrl}/${id}`);
  }

  async updateEmployee(
    id: number,
    employeeData: UpdateEmployeeRequest
  ): Promise<ApiResponse<Employee>> {
    return apiClient.put(`${this.baseUrl}/${id}`, employeeData);
  }

  async deleteEmployee(id: number): Promise<ApiResponse<void>> {
    return apiClient.delete(`${this.baseUrl}/${id}`);
  }

  async activateEmployee(id: number): Promise<ApiResponse<Employee>> {
    return apiClient.patch(`${this.baseUrl}/${id}/activate`);
  }

  async deactivateEmployee(id: number): Promise<ApiResponse<Employee>> {
    return apiClient.patch(`${this.baseUrl}/${id}/deactivate`);
  }

  async updateEmployeeRole(
    id: number,
    newRole: number
  ): Promise<ApiResponse<Employee>> {
    return apiClient.put(`${this.baseUrl}/${id}/role?newRole=${newRole}`);
  }
}

export const employeeService = new EmployeeService();
export default employeeService;
